### PR TITLE
Fix getting live application routes on bg-deploy

### PR DESCRIPTION
**Pull request by theghost5800**
Also fix determine of live application color based on previous operation
variables

JIRA:LMCROSSITXSADEPLOY-1875

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculatorTest.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+               Arguments.of(Arrays.asList(MapUtil.asMap("module-1", "foo"), MapUtil.asMap("module-1", "foo-blue"), MapUtil.asMap("module-1", "foo-green")), ApplicationColor.GREEN, 
+                            MapUtil.asMap("module-1", "foo-green"))
+        // @formatter:on
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    public void testGetLiveDeployedModule(List<Map<String, String>> deployedModulesNames, ApplicationColor deployedMtaColor,
+                                          Map<String, String> expectedLiveDeployedModuleName) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculatorTest.java]: Why do you need map here(expectedLiveDeployedModuleName), only the module names?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java [hunk 2]
```diff
@@ -26,8 +27,6 @@
 @Named("applicationColorDetector")
 public class ApplicationColorDetector {
 
-    private static final ApplicationColor COLOR_OF_APPLICATIONS_WITHOUT_SUFFIX = ApplicationColor.BLUE;
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: Why this is removed from here?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java [hunk 3]
```diff
@@ -69,7 +68,7 @@ public ApplicationColor detectLiveApplicationColor(DeployedMta deployedMta, Stri
         if (latestDeployedColor == null) {
             return olderApplicationColor;
         }
-        return phase == Phase.UNDEPLOY ? latestDeployedColor : olderApplicationColor;
+        return phase == Phase.UNDEPLOY ? latestDeployedColor : latestDeployedColor.getAlternativeColor();
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: This should be in a separate commit, because it addresses a different issue.

Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: done

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java [hunk 4]
```diff
@@ -104,7 +103,7 @@ private ApplicationColor getApplicationColor(DeployedMtaModule deployedMtaModule
                      .filter(color -> deployedMtaModule.getAppName()
                                                        .endsWith(color.asSuffix()))
                      .findFirst()
-                     .orElse(COLOR_OF_APPLICATIONS_WITHOUT_SUFFIX);
+                     .orElse(DeployedLiveModuleCalculator.COLOR_OF_APPLICATIONS_WITHOUT_SUFFIX);
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: I think this constant is better here, rather than in the calculator

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java [hunk 5]
```diff
@@ -58,19 +58,20 @@
...
     protected final ApplicationEnvironmentCloudModelBuilder applicationEnvCloudModelBuilder;
-    protected final DeployedMta deployedMta;
+    protected final DeployedLiveModuleCalculator deployedLiveModuleCalculator;
     protected final UserMessageLogger stepLogger;
 
     protected final ParametersChainBuilder parametersChainBuilder;
 
-    public ApplicationCloudModelBuilder(DeploymentDescriptor deploymentDescriptor, boolean prettyPrinting, DeployedMta deployedMta,
-                                        String deployId, UserMessageLogger stepLogger) {
+    public ApplicationCloudModelBuilder(DeploymentDescriptor deploymentDescriptor, boolean prettyPrinting,
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: Add a builder.
This many arguments is a bit much

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculator.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+public class DeployedLiveModuleCalculator {
+
+    public static final ApplicationColor COLOR_OF_APPLICATIONS_WITHOUT_SUFFIX = ApplicationColor.BLUE;
+
+    private DeployedMta deployedMta;
+
+    private ApplicationColor deployedMtaColor;
+
+    public DeployedLiveModuleCalculator(DeployedMta deployedMta, ApplicationColor deployedMtaColor) {
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculator.java]: Maybe a better name would be `LiveDeployedModuleCalculator`

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java (hunk 4 above)
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: The reason for moving is that DeployedModuleCalculator doesn't have dependency to cf.process package and can't access classes from cf.processes, so the constant should be moved somewhere where it could be accessible from cf.core and cf.processes packages.

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculatorTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        DeployedMta deployedMta = buildDeployedMta(deployedModulesNames);
+        DeployedMtaModule expectedDeployedModule = buildDeployedModule(expectedLiveDeployedModuleName);
+        String moduleName = expectedDeployedModule != null ? expectedDeployedModule.getModuleName() : "";
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculatorTest.java]: Yes, in case when you don't have deployed modules and tested method should return null

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculator.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        return deployedModules.isEmpty() ? null : deployedModules.get(0);
+    }
+
+    private List<DeployedMtaModule> getDeployedModules() {
+        return deployedMta.getModules();
+    }
+
+    private boolean isModuleNamesEqual(DeployedMtaModule deployedModule, String moduleName) {
+        return deployedModule.getModuleName()
+                             .equalsIgnoreCase(moduleName);
```
[the remark this answers is not in the document]
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedLiveModuleCalculator.java]: I am using the same logic from DeployedMta class for find module by name:

`public DeployedMtaModule findDeployedModule(String moduleName) {	
        return getModules().stream()	
                           .filter(module -> module.getModuleName()	
                                                   .equalsIgnoreCase(moduleName))	
                           .findFirst()	
                           .orElse(null);	
    }`

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java (hunk 2 above)
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/helpers/ApplicationColorDetector.java]: It was moved from cf.process in cf.core package because the constant couldn't be used from classes in cf.core package.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java (hunk 5 above)
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: I tried to use Immutables for builder but it will be need more time to refactor due to inheritence from V3 application cloud model builder.

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v2/CloudModelBuilderTest.java [hunk 9]
```diff
@@ -23,7 +23,7 @@
 import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudApplicationExtended;
 import com.sap.cloud.lm.sl.cf.client.lib.domain.CloudServiceExtended;
 import com.sap.cloud.lm.sl.cf.core.cf.HandlerFactory;
-import com.sap.cloud.lm.sl.cf.core.cf.util.DeployedLiveModuleCalculator;
+import com.sap.cloud.lm.sl.cf.core.cf.util.DeployedModuleCalculatorColor;
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v2/CloudModelBuilderTest.java]: The change to `DeployedModuleCalculator` is good but with the `Color` suffix, not so much.
My suggestion is name the concrete class something like `DeployedModuleCalculatorWithColor`

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java (hunk 5 above)
Comment (radito3) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: Then write a custom builder

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                             .equalsIgnoreCase(moduleName);
+    }
+
+    private boolean hasDeployedModuleContainLiveColor(DeployedMtaModule deployedModule) {
+        String appName = deployedModule.getAppName();
+        return appName.endsWith(deployedMtaColor.asSuffix())
+            || (hasDeployedModuleContainColorWithoutSuffix(appName) && doesDeployedColorMatchColorWithoutSuffix());
+    }
+
+    private boolean hasDeployedModuleContainColorWithoutSuffix(String appName) {
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java]: `doesAppNameHaveColorSuffix`

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    private List<DeployedMtaModule> getDeployedModules() {
+        return deployedMta.getModules();
+    }
+
+    private boolean isModuleNameMatching(DeployedMtaModule deployedModule, String moduleName) {
+        return deployedModule.getModuleName()
+                             .equalsIgnoreCase(moduleName);
+    }
+
+    private boolean hasDeployedModuleContainLiveColor(DeployedMtaModule deployedModule) {
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java]: `doesDeployedModuleContainLiveColor`

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+            return deployedModules.filter(this::hasDeployedModuleContainLiveColor)
+                                  .findFirst()
+                                  .orElse(null);
+        }
+
+        return deployedModules.findFirst()
+                              .orElse(null);
+    }
+
+    private List<DeployedMtaModule> getDeployedModules() {
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/util/DeployedModuleCalculatorColor.java]: I don't think the amount of uses and length (as an expression) of `deployedMta.getModules` warrants the need for this method. Just use it normally and delete this

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculatorWithColor.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public List<DeployedMtaModule> updateIdleDeployedModules() {
+        return deployedModules.stream()
+                              .map(this::setIdleModules)
+                              .collect(Collectors.toList());
+    }
+
+    private DeployedMtaModule setIdleModules(DeployedMtaModule deployedModule) {
+        if (doesDeployedModuleNotContainLiveColor(deployedModule)) {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculatorWithColor.java]: Could we rename this method to doesDeployedModuleHaveIdleColor?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculator.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package com.sap.cloud.lm.sl.cf.process.util;
+
+import java.util.List;
+
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMtaModule;
+
+public interface DeployedModuleCalculator {
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculator.java]: Calculator is used for calculating one thing from two or more other things.
Your class is taking one set of modules and it is returning the same set.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculator.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+
+    List<DeployedMtaModule> updateIdleDeployedModules();
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculator.java]: Think of a better naming for this method

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculatorWithColor.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    private boolean doesDeployedModuleNotContainLiveColor(DeployedMtaModule deployedModule) {
+        String appName = deployedModule.getAppName();
+        return !appName.endsWith(deployedMtaColor.asSuffix())
+            && !(doesDeployedModuleContainColorWithoutSuffix(appName) && doesDeployedColorMatchColorWithoutSuffix());
+    }
+
+    private boolean doesDeployedModuleContainColorWithoutSuffix(String appName) {
+        return !appName.endsWith(ApplicationColor.BLUE.asSuffix()) && !appName.endsWith(ApplicationColor.GREEN.asSuffix());
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DeployedModuleCalculatorWithColor.java]: Is this logic extensible?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java [hunk 17]
```diff
@@ -418,6 +418,7 @@
     public static final String ARCHIVE_IS_VERIFIED = "Archive signature is verified";
     public static final String CHECKING_FOR_OVERWRITING_READ_ONLY_PARAMETERS = "Checking for overwriting read-only parameters for mta with id: \"{0}\"";
     public static final String NO_READ_ONLY_PARAMETERS_ARE_OVERWRITTEN = "No read-only parameters are overwritten";
+    public static final String MODULE_WITH_APPLICATION_NAME_WAS_MARKED = "Module \"{0}\" with application name \"{1}\" was marked as idle";
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/message/Messages.java]: The constant name suggests more generic message than the actual message is.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java [hunk 18]
```diff
@@ -45,9 +49,12 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
             ApplicationColor liveMtaColor = getLiveApplicationColor(deployedMta, execution);
             ApplicationColor idleMtaColor = liveMtaColor.getAlternativeColor();
             getStepLogger().info(Messages.ASSUMED_LIVE_AND_IDLE_COLORS, liveMtaColor, idleMtaColor);
+            deployedMtaColor = liveMtaColor;
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java]: I need some explanation about this.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java [hunk 19]
```diff
@@ -68,6 +75,16 @@ private ApplicationColor getLiveApplicationColor(DeployedMta deployedMta, Execut
                                                                                                   .getVariable(Constants.VAR_CORRELATION_ID));
     }
 
+    private void setIdleModules(ExecutionWrapper execution, DeployedMta deployedMta, ApplicationColor deployedMtaColor) {
+        if (deployedMta != null && deployedMtaColor != null) {
+            List<DeployedMtaModule> updatedModules = new DetermineDeployedModuleLabelBasedOnColor(deployedMta.getModules(),
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java]: This.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
+package com.sap.cloud.lm.sl.cf.process.util;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMtaModule;
+import com.sap.cloud.lm.sl.cf.core.model.DeployedModuleLabel;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+
+public abstract class DetermineDeployedModuleLabel {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java]: Using a verb for a class name.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    private List<DeployedMtaModule> deployedModules;
+
+    private StepLogger stepLogger;
+
+    public DetermineDeployedModuleLabel(List<DeployedMtaModule> deployedModules, StepLogger stepLogger) {
+        this.deployedModules = deployedModules;
+        this.stepLogger = stepLogger;
+    }
+
+    public List<DeployedMtaModule> updateIdleDeployedModules() {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java]: The name of this method does not fit with the class.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+        return deployedModules.stream()
+                              .map(this::setIdleModules)
+                              .collect(Collectors.toList());
+    }
+
+    private DeployedMtaModule setIdleModules(DeployedMtaModule deployedModule) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/DetermineDeployedModuleLabel.java]: Setters should not return values.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java [hunk 23]
```diff
@@ -128,16 +128,17 @@ private ApplicationUrisCloudModelBuilder getApplicationUrisCloudModelBuilder(Lis
...
-        DeployedMtaModule deployedModule = findDeployedModule(deployedMta, module);
+        DeployedMtaApplication deployedModule = findDeployedMtaModule(module);
         return getApplicationUrisCloudModelBuilder(parametersList).getApplicationUris(module, parametersList, deployedModule);
     }
 
-    protected <R> R parseParameters(List<Map<String, Object>> parametersList, ParametersParser<R> parser) {
-        return parser.parse(parametersList);
+    private DeployedMtaApplication findDeployedMtaModule(Module module) {
+        return deployedMta == null ? null
+            : deployedMta.findLiveApplication(module.getName(), DeployedMtaApplication.ProductizationState.LIVE);
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v2/ApplicationCloudModelBuilder.java]: Remove `Live` from the name of the method.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/DeployedMta.java [hunk 24]
```diff
@@ -63,12 +65,17 @@ public boolean equals(Object object) {
...
         return getModules().stream()
-                           .filter(module -> module.getModuleName()
-                                                   .equalsIgnoreCase(moduleName))
+                           .filter(deployedApplication -> doesDeployedApplicationMatch(moduleName, deployedApplication, state))
                            .findFirst()
                            .orElse(null);
     }
 
+    private boolean doesDeployedApplicationMatch(String moduleName, DeployedMtaApplication deployedApplication, ProductizationState state) {
+        return deployedApplication.getModuleName()
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/DeployedMta.java]: These are two separate checks. Use two filter() calls in the method above.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdaterColor.java [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+    public ApplicationProductizationStateUpdaterColor(List<DeployedMtaApplication> deployedApplications, ApplicationColor liveMtaColor,
+                                                    StepLogger stepLogger) {
+        super(deployedApplications, stepLogger);
+        this.liveMtaColor = liveMtaColor;
+    }
+
+    @Override
+    protected boolean doesDeployedApplicationHasIdleLabel(DeployedMtaApplication deployedApplication) {
+        String appName = deployedApplication.getAppName();
+        return !(appName.endsWith(liveMtaColor.asSuffix())
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdaterColor.java]: Why do we still need this check?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdaterColor.java [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+            || CloudModelBuilderUtil.getApplicationColor(deployedApplication) == liveMtaColor);
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdaterColor.java]: Shouldn't this be negated?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java [hunk 27]
```diff
@@ -68,6 +70,16 @@ private ApplicationColor getLiveApplicationColor(DeployedMta deployedMta, Execut
                                                                                                   .getVariable(Constants.VAR_CORRELATION_ID));
     }
 
+    private void setIdleModules(ExecutionWrapper execution, DeployedMta deployedMta, ApplicationColor liveMtaColor) {
+        if (deployedMta != null && liveMtaColor != null) {
+            List<DeployedMtaApplication> updatedModules = new ApplicationProductizationStateUpdaterColor(deployedMta.getModules(),
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BlueGreenRenameStep.java]: Why do we need `Color` at the end of the class name?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdater.java [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.sap.cloud.lm.sl.cf.core.model.DeployedMtaApplication;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+
+public abstract class ApplicationProductizationStateUpdater {
+
+    private List<DeployedMtaApplication> deployedModules;
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdater.java]: This state is only used in the `updateIdleDeployedApplications` method. Why not pass it as an argument to the method itself?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdater.java [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+
+    private StepLogger stepLogger;
+
+    public ApplicationProductizationStateUpdater(List<DeployedMtaApplication> deployedModules, StepLogger stepLogger) {
+        this.deployedModules = deployedModules;
+        this.stepLogger = stepLogger;
+    }
+
+    public List<DeployedMtaApplication> updateIdleDeployedApplications() {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/ApplicationProductizationStateUpdater.java]: The name of this method still does not match the name of the class.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/DeployedMta.java (hunk 24 above)
Comment (theghost5800) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/DeployedMta.java]: fixed
